.github/workflows/matrix.yml

```
@@ -16,72 +16,124 @@ on:
+      should_skip_code: ${{ steps.skip_check.outputs.should_skip }}
     steps:
       - id: skip_check
         uses: fkirc/skip-duplicate-actions@master
         with:
-          cancel_others: "true"
-          paths_ignore: '["android/**", "build-data/osx/**", "doc/**", "doxygen_doc/**", "lgtm/**", "msvc-**", "object_creator/**", "tools/**", "utilities/**"]'
+          paths_ignore: '["android/**", "build-data/osx/**", "doc/**", "doxygen_doc/**", "msvc-**", "object_creator/**", "tools/**", "utilities/**", "scripts/**", "data/**", "lang/**"]'
+
+  skip-duplicates-data:
+    continue-on-error: true
+    runs-on: ubuntu-22.04
+    # Map a step output to a job output
+    outputs:
+      should_skip_data: ${{ steps.skip_check.outputs.should_skip }}
+    steps:
+      - id: skip_check
+        uses: fkirc/skip-duplicate-actions@master
+        with:
+          paths_ignore: '["android/**", "build-data/osx/**", "doc/**", "doxygen_doc/**", "msvc-**", "object_creator/**", "tools/**", "utilities/**", "scripts/**"]'
+
+  matrix-variables:
+    permissions:
+      contents: none
+    continue-on-error: true
+    runs-on: ubuntu-latest
+    outputs:
+      fail_fast: ${{ steps.matrix_vars.outputs.fail_fast }}
+      skip_tests: ${{ steps.matrix_vars.outputs.skip_tests }}
+      max_parallel: ${{ steps.matrix_vars.outputs.max_parallel }}
+    steps:
+      - id: matrix_vars
+        run: |
+          echo "fail_fast=$([ "$GITHUB_REF_NAME" = "upload" ] && echo false || echo true)" >> $GITHUB_OUTPUT
+          echo "skip_tests=$([ "$GITHUB_REF_NAME" = "upload" ] && echo true || echo false)" >> $GITHUB_OUTPUT
+          echo "max_parallel=$([ "$GITHUB_REF_NAME" = "upload" ] && echo 20 || echo 1)" >> $GITHUB_OUTPUT
+
   varied_builds:
-    needs: skip-duplicates
+    needs: [ skip-duplicates-code, skip-duplicates-data, matrix-variables ]
     strategy:
+      fail-fast: ${{ fromJSON(needs.matrix-variables.outputs.fail_fast) }}
+      max-parallel: ${{ fromJSON(needs.matrix-variables.outputs.max_parallel) }}
       matrix:
         include:
-          - compiler: g++-12
+          - title: GCC 12, Ubuntu, Curses
+            compiler: g++-12
             os: ubuntu-22.04
             cmake: 0
             tiles: 0
             sound: 0
             lua: 0
             test-stage: 1
             libbacktrace: 1
-            title: GCC 12, Ubuntu, Curses
             native: linux64
-          - compiler: g++-12
+            ccache_limit: 2G
+            ccache_key: linux-gcc-12
+
+          - title: GCC 12, Ubuntu, Tiles, Sound, Lua, CMake, Languages
+            compiler: g++-12
             os: ubuntu-22.04
             cmake: 1
             tiles: 1
             sound: 1
             lua: 1
             languages: all
             native: linux64
-            title: GCC 12, Ubuntu, Tiles, Sound, Lua, CMake, Languages
-          - compiler: g++-12
+            ccache_limit: 4G
+            ccache_key: linux-gcc-12-cmake-lua
```

Tiles/no tiles has more impact than Lua
```suggestion
            ccache_key: linux-gcc-12-cmake-tiles
```